File: iOS_macOS/starflower/starflower/StargazingViewModel.swift
//
//  StargazingViewModel.swift
//  starflower
//
//  Created by 양지성 on 6/19/26.
//

import Foundation
import SwiftUI
import Combine
import WidgetKit

@MainActor
final class StargazingViewModel: ObservableObject {
    @Published var data: StargazingData?
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var showSearch = false
    @Published var savedLocation: SavedLocation? { didSet { persist() } }

    private let weather = WeatherService.shared
    private let appGroup = UserDefaults(suiteName: "group.com.songsari.starflower")
    private let key = "starflower.location.v1"

    private let weekdays = ["일","월","화","수","목","금","토"]
    private var cal: Calendar { Calendar.current }

    init() {
        if let d = appGroup?.data(forKey: key),
           let loc = try? JSONDecoder().decode(SavedLocation.self, from: d) {
            savedLocation = loc
        } else {
            showSearch = true
        }
    }

    func selectLocation(_ r: GeoResult) {
        savedLocation = SavedLocation(name: r.name, admin1: r.admin1, country: r.country,
                                      latitude: r.latitude, longitude: r.longitude)
        showSearch = false
        data = nil           // 이전 데이터 치워서 로딩창 표시
        Task { await loadData() }
    }

    func loadData() async {
        guard let loc = savedLocation else { return }
        isLoading = true; errorMessage = nil
        do {
            async let w = weather.fetchWeather(lat: loc.latitude, lng: loc.longitude)
            async let a = weather.fetchAir(lat: loc.latitude, lng: loc.longitude)
            let (wx, air) = try await (w, a)
            data = buildData(loc, wx, air)
            saveForWidget(loc)
        } catch {
            errorMessage = "날씨 정보를 불러오지 못했어요."
        }
        isLoading = false
    }

    private func persist() {
        guard let loc = savedLocation, let e = try? JSONEncoder().encode(loc) else { return }
        appGroup?.set(e, forKey: key)
    }
    private func saveForWidget(_ loc: SavedLocation) {
        appGroup?.set(loc.latitude, forKey: "latitude")
        appGroup?.set(loc.longitude, forKey: "longitude")
        appGroup?.set(loc.name, forKey: "locationName")
        WidgetCenter.shared.reloadAllTimelines()
    }

    // ── 데이터 조립 (웹 buildData 동일) ───────────────────
    private func buildData(_ loc: SavedLocation, _ wx: WeatherResponse, _ air: AirResponse?) -> StargazingData {
        let now = Date()
        let cur = wx.current
        let hourly = wx.hourly
        let lat = loc.latitude, lng = loc.longitude

        let hour = cal.component(.hour, from: now)
        let todayEvening = hour < 6 ? cal.date(byAdding: .day, value: -1, to: now)! : now

        var forecast: [DayForecast] = []
        var n0Cloud = cur.cloudCover, n0Hum = cur.relativeHumidity2m
        var n0Wind = cur.windSpeed10m, n0Pm = 0.0

        for i in 0..<3 {
            let evening = cal.date(byAdding: .day, value: i, to: todayEvening)!
            let eStr = dateStr(evening)
            let mStr = dateStr(cal.date(byAdding: .day, value: 1, to: evening)!)
            let idxs = nightIndices(hourly.time, eStr, mStr)

            let cloud = avg(idxs.map { hourly.cloudCover[$0] })
            let hum   = avg(idxs.map { hourly.relativeHumidity2m[$0] })
            let wind  = avg(idxs.map { hourly.windSpeed10m[$0] })
            let pm    = airAvg(air, hourly.time, idxs)
            if i == 0 { n0Cloud = cloud; n0Hum = hum; n0Wind = wind; n0Pm = pm }

            let moonMid = MoonCalculator.illumination(date: nightMidnight(eStr))
            let exposure = MoonCalculator.nightExposure(
                start: nightStart(eStr), end: nightEnd(eStr), lat: lat, lng: lng)

            let score = ScoreCalculator.compute(NightInputs(
                cloud: cloud, humidity: hum, pm25: pm, wind: wind,
                moonIllum: moonMid.fraction, moonExposure: exposure))

            let repIdx = idxs.first { hourly.time[$0].hasSuffix("T23:00") } ?? idxs.first ?? -1
            let repCode = repIdx >= 0 ? hourly.weatherCode[repIdx] : cur.weatherCode
            let cond = classifySky(repCode, cloud)

            let label = i == 0 ? "오늘" : weekdays[cal.component(.weekday, from: evening) - 1]
            forecast.append(DayForecast(date: evening, label: label, score: score, condition: cond))
        }

        let moonNow = MoonCalculator.illumination(date: now)
        let moonPos = MoonCalculator.position(date: now, lat: lat, lng: lng)
        let condition = forecast.first?.condition ?? classifySky(cur.weatherCode, cur.cloudCover)

        // 낮/밤/노을
        let sunNow = MoonCalculator.getSunAltitude(date: now, lat: lat, lng: lng)
        let sunLater = MoonCalculator.getSunAltitude(date: now.addingTimeInterval(600), lat: lat, lng: lng)
        let tw = 0.12
        let daypart: Daypart
        if sunNow > tw { daypart = .day }
        else if sunNow < -tw { daypart = .night }
        else { daypart = sunLater > sunNow ? .dawn : .dusk }
        
        // past_days=1 추가로 daily[0]=어제, [1]=오늘, [2]=내일
        let daily = wx.daily
        let todayIdx = 1  // 오늘은 항상 index 1
        let isEarly = hour < 6

        // 0~6시: 전일 일몰(daily[0]) + 금일 일출(daily[1])
        // 나머지: 금일 일몰(daily[1]) + 익일 일출(daily[2])
        let chosenSunset  = parseDate(daily.sunset[isEarly ? 0 : todayIdx])
        let chosenSunrise = parseDate(daily.sunrise[isEarly ? todayIdx : min(todayIdx + 1, daily.sunrise.count - 1)])

        // 0~6시엔 전일 기준, 나머지는 오늘 기준
        let moonRefDate: Date
        if hour < 6 {
            moonRefDate = cal.date(byAdding: .day, value: -1, to: cal.startOfDay(for: now))!
        } else {
            moonRefDate = cal.startOfDay(for: now)
        }
        let moonRS = MoonCalculator.moonRiseSet(
            referenceDate: moonRefDate, lat: lat, lng: lng)
        
        return StargazingData(
            location: loc,
            score: forecast.first?.score ?? 0,
            condition: condition,
            daypart: daypart,
            temperature: cur.temperature2m,
            pressure: cur.surfacePressure,
            nightCloud: n0Cloud, nightHumidity: n0Hum, nightWind: n0Wind, nightPm25: n0Pm,
            sunrise: chosenSunrise,
            sunset: chosenSunset,
            moonIllum: moonNow.fraction, moonPhase: moonNow.phase,
            moonAltitude: moonPos.altitude,
            moonName: ScoreCalculator.moonPhaseName(phase: moonNow.phase),
            moonrise: moonRS.rise,
            moonset: moonRS.set,
            forecast: forecast,
            updatedAt: now)
    }

    // ── 유틸 ──────────────────────────────────────────────
    private func dateStr(_ d: Date) -> String {
        let f = DateFormatter(); f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "yyyy-MM-dd"; return f.string(from: d)
    }
    private func nightIndices(_ times: [String], _ e: String, _ m: String) -> [Int] {
        times.enumerated().compactMap { i, t in
            let d = String(t.prefix(10))
            guard let h = Int(t.dropFirst(11).prefix(2)) else { return nil }
            if d == e && h >= 19 { return i }
            if d == m && h <= 5  { return i }
            return nil
        }
    }
    private func ymd(_ s: String) -> DateComponents? {
        let p = s.split(separator: "-").compactMap { Int($0) }
        guard p.count == 3 else { return nil }
        return DateComponents(year: p[0], month: p[1], day: p[2])
    }
    private func nightStart(_ e: String) -> Date {
        var c = ymd(e)!; c.hour = 19; return cal.date(from: c) ?? Date()
    }
    private func nightMidnight(_ e: String) -> Date {
        var c = ymd(e)!; c.hour = 24; return cal.date(from: c) ?? Date()  // 다음날 00:00
    }
    private func nightEnd(_ e: String) -> Date {
        var c = ymd(e)!; c.day! += 1; c.hour = 5; return cal.date(from: c) ?? Date()
    }
    private func avg(_ v: [Double]) -> Double {
        let f = v.filter { $0.isFinite }
        return f.isEmpty ? 0 : f.reduce(0, +) / Double(f.count)
    }
    private func airAvg(_ air: AirResponse?, _ times: [String], _ idxs: [Int]) -> Double {
        guard let air else { return 0 }
        let vals: [Double] = idxs.compactMap { i in
            guard i < times.count, let j = air.hourly.time.firstIndex(of: times[i]) else { return nil }
            return air.hourly.pm25[j]
        }
        return avg(vals)
    }
    private func classifySky(_ code: Int, _ cloud: Double) -> SkyCondition {
        if code >= 95 { return .rain }
        if (71...77).contains(code) || code == 85 || code == 86 { return .snow }
        if (51...67).contains(code) || (80...82).contains(code) { return .rain }
        if code == 45 || code == 48 { return .fog }
        if cloud < 15 { return .clear }
        if cloud < 50 { return .partly }
        if cloud < 85 { return .cloudy }
        return .overcast
    }
    private func parseDate(_ s: String) -> Date {
        let f = DateFormatter(); f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = .current
        f.dateFormat = "yyyy-MM-dd'T'HH:mm"
        return f.date(from: s) ?? Date()
    }
}

File: Android/starflower/app/src/main/java/com/songsari/starflower/ui/StargazingViewModel.kt
package com.songsari.starflower.ui

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.songsari.starflower.calc.MoonCalculator
import com.songsari.starflower.calc.ScoreCalculator
import com.songsari.starflower.data.AirResponse
import com.songsari.starflower.data.LocationStore
import com.songsari.starflower.data.WeatherResponse
import com.songsari.starflower.data.WeatherService
import com.songsari.starflower.model.DayForecast
import com.songsari.starflower.model.Daypart
import com.songsari.starflower.model.GeoResult
import com.songsari.starflower.model.NightInputs
import com.songsari.starflower.model.SavedLocation
import com.songsari.starflower.model.SkyCondition
import com.songsari.starflower.model.StargazingData
import com.songsari.starflower.widget.WidgetScheduler
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone

class StargazingViewModel(app: Application) : AndroidViewModel(app) {

    private val _data = MutableStateFlow<StargazingData?>(null)
    val data: StateFlow<StargazingData?> = _data.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _showSearch = MutableStateFlow(false)
    val showSearch: StateFlow<Boolean> = _showSearch.asStateFlow()

    private val _savedLocation = MutableStateFlow<SavedLocation?>(null)
    val savedLocation: StateFlow<SavedLocation?> = _savedLocation.asStateFlow()

    private val weekdays = listOf("일", "월", "화", "수", "목", "금", "토")
    private fun cal() = Calendar.getInstance(TimeZone.getDefault())

    init {
        viewModelScope.launch {
            val loc = LocationStore.load(getApplication())
            if (loc != null) {
                _savedLocation.value = loc
                loadData()
            } else {
                _showSearch.value = true
            }
        }
    }

    fun setShowSearch(v: Boolean) { _showSearch.value = v }

    fun selectLocation(r: GeoResult) {
        val loc = SavedLocation(r.name, r.admin1, r.country, r.latitude, r.longitude)
        _savedLocation.value = loc
        _showSearch.value = false
        _data.value = null   // 이전 데이터 치워 로딩창 표시
        viewModelScope.launch {
            LocationStore.save(getApplication(), loc)
            WidgetScheduler.updateNow(getApplication())
            loadData()
        }
    }

    fun loadData() {
        val loc = _savedLocation.value ?: return
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            try {
                val wxDeferred = async { WeatherService.fetchWeather(loc.latitude, loc.longitude) }
                val airDeferred = async { WeatherService.fetchAir(loc.latitude, loc.longitude) }
                val wx = wxDeferred.await()
                val air = airDeferred.await()
                _data.value = buildData(loc, wx, air)
                // 데이터 갱신 후
                WidgetScheduler.updateNow(getApplication())
            } catch (e: Exception) {
                _errorMessage.value = "날씨 정보를 불러오지 못했어요."
            } finally {
                _isLoading.value = false
            }
        }
    }

    // ── 데이터 조립 (iOS buildData 1:1) ───────────────────
    private fun buildData(loc: SavedLocation, wx: WeatherResponse, air: AirResponse?): StargazingData {
        val now = Date()
        val cur = wx.current
        val hourly = wx.hourly
        val lat = loc.latitude
        val lng = loc.longitude

        val c = cal().apply { time = now }
        val hour = c.get(Calendar.HOUR_OF_DAY)
        val todayEvening = if (hour < 6) addDays(now, -1) else now

        val forecast = ArrayList<DayForecast>()
        var n0Cloud = cur.cloudCover
        var n0Hum = cur.relativeHumidity2m
        var n0Wind = cur.windSpeed10m
        var n0Pm = 0.0

        for (i in 0 until 3) {
            val evening = addDays(todayEvening, i)
            val eStr = dateStr(evening)
            val mStr = dateStr(addDays(evening, 1))
            val idxs = nightIndices(hourly.time, eStr, mStr)

            val cloud = avg(idxs.map { hourly.cloudCover[it] })
            val hum = avg(idxs.map { hourly.relativeHumidity2m[it] })
            val wind = avg(idxs.map { hourly.windSpeed10m[it] })
            val pm = airAvg(air, hourly.time, idxs)
            if (i == 0) { n0Cloud = cloud; n0Hum = hum; n0Wind = wind; n0Pm = pm }

            val moonMid = MoonCalculator.illumination(nightMidnight(eStr))
            val exposure = MoonCalculator.nightExposure(nightStart(eStr), nightEnd(eStr), lat, lng)
            val score = ScoreCalculator.compute(
                NightInputs(cloud, hum, pm, wind, moonMid.fraction, exposure)
            )
            val repIdx = idxs.firstOrNull { hourly.time[it].endsWith("T23:00") } ?: idxs.firstOrNull() ?: -1
            val repCode = if (repIdx >= 0) hourly.weatherCode[repIdx] else cur.weatherCode
            val cond = classifySky(repCode, cloud)
            val label = if (i == 0) "오늘" else weekdays[weekdayIndex(evening)]
            forecast.add(DayForecast(evening, label, score, cond))
        }

        val moonNow = MoonCalculator.illumination(now)
        val moonPos = MoonCalculator.position(now, lat, lng)
        val condition = forecast.firstOrNull()?.condition ?: classifySky(cur.weatherCode, cur.cloudCover)

        // 낮/밤/노을
        val sunNow = MoonCalculator.getSunAltitude(now, lat, lng)
        val sunLater = MoonCalculator.getSunAltitude(Date(now.time + 600_000), lat, lng)
        val tw = 0.12
        val daypart = when {
            sunNow > tw -> Daypart.DAY
            sunNow < -tw -> Daypart.NIGHT
            else -> if (sunLater > sunNow) Daypart.DAWN else Daypart.DUSK
        }

        // past_days=1 → daily[0]=어제, [1]=오늘, [2]=내일
        val daily = wx.daily
        val todayIdx = 1
        val isEarly = hour < 6
        val chosenSunset = parseDate(daily.sunset[if (isEarly) 0 else todayIdx])
        val chosenSunrise = parseDate(
            daily.sunrise[if (isEarly) todayIdx else minOf(todayIdx + 1, daily.sunrise.size - 1)]
        )

        // 월출·월몰: 0~6시면 전일 기준, 그 외 오늘 기준
        val moonRefDate: Date = if (hour < 6) startOfDay(addDays(now, -1)) else startOfDay(now)
        val moonRS = MoonCalculator.moonRiseSet(moonRefDate, lat, lng)

        return StargazingData(
            location = loc,
            score = forecast.firstOrNull()?.score ?: 0,
            condition = condition,
            daypart = daypart,
            temperature = cur.temperature2m,
            pressure = cur.surfacePressure,
            nightCloud = n0Cloud,
            nightHumidity = n0Hum,
            nightWind = n0Wind,
            nightPm25 = n0Pm,
            sunrise = chosenSunrise,
            sunset = chosenSunset,
            moonIllum = moonNow.fraction,
            moonPhase = moonNow.phase,
            moonAltitude = moonPos.altitude,
            moonName = ScoreCalculator.moonPhaseName(moonNow.phase),
            moonrise = moonRS.rise,
            moonset = moonRS.set,
            forecast = forecast,
            updatedAt = now,
        )
    }

    // ── 유틸 ──────────────────────────────────────────────
    private val ymdFmt = SimpleDateFormat("yyyy-MM-dd", Locale.US).apply {
        timeZone = TimeZone.getDefault()
    }
    private val isoFmt = SimpleDateFormat("yyyy-MM-dd'T'HH:mm", Locale.US).apply {
        timeZone = TimeZone.getDefault()
    }

    private fun dateStr(d: Date): String = ymdFmt.format(d)

    private fun addDays(d: Date, n: Int): Date {
        val c = cal().apply { time = d }
        c.add(Calendar.DAY_OF_MONTH, n)
        return c.time
    }

    private fun startOfDay(d: Date): Date {
        val c = cal().apply { time = d }
        c.set(Calendar.HOUR_OF_DAY, 0); c.set(Calendar.MINUTE, 0)
        c.set(Calendar.SECOND, 0); c.set(Calendar.MILLISECOND, 0)
        return c.time
    }

    private fun weekdayIndex(d: Date): Int {
        val c = cal().apply { time = d }
        return c.get(Calendar.DAY_OF_WEEK) - 1   // 일=1 → 0
    }

    private fun nightIndices(times: List<String>, e: String, m: String): List<Int> =
        times.indices.filter { i ->
            val t = times[i]
            val d = t.take(10)
            val h = t.drop(11).take(2).toIntOrNull() ?: return@filter false
            (d == e && h >= 19) || (d == m && h <= 5)
        }

    private fun ymdToDate(s: String, hour: Int, addDay: Int = 0): Date {
        val parts = s.split("-").mapNotNull { it.toIntOrNull() }
        val c = cal()
        c.clear()
        c.set(parts[0], parts[1] - 1, parts[2] + addDay, hour, 0, 0)
        return c.time
    }

    private fun nightStart(e: String) = ymdToDate(e, 19)
    private fun nightMidnight(e: String) = ymdToDate(e, 24)      // 다음날 00:00
    private fun nightEnd(e: String) = ymdToDate(e, 5, addDay = 1)

    private fun avg(v: List<Double>): Double {
        val f = v.filter { it.isFinite() }
        return if (f.isEmpty()) 0.0 else f.sum() / f.size
    }

    private fun airAvg(air: AirResponse?, times: List<String>, idxs: List<Int>): Double {
        if (air == null) return 0.0
        val vals = idxs.mapNotNull { i ->
            if (i >= times.size) return@mapNotNull null
            val j = air.hourly.time.indexOf(times[i])
            if (j >= 0) air.hourly.pm25.getOrNull(j) else null
        }
        return avg(vals)
    }

    private fun classifySky(code: Int, cloud: Double): SkyCondition = when {
        code >= 95 -> SkyCondition.RAIN
        code in 71..77 || code == 85 || code == 86 -> SkyCondition.SNOW
        code in 51..67 || code in 80..82 -> SkyCondition.RAIN
        code == 45 || code == 48 -> SkyCondition.FOG
        cloud < 15 -> SkyCondition.CLEAR
        cloud < 50 -> SkyCondition.PARTLY
        cloud < 85 -> SkyCondition.CLOUDY
        else -> SkyCondition.OVERCAST
    }

    private fun parseDate(s: String): Date = try {
        isoFmt.parse(s) ?: Date()
    } catch (e: Exception) {
        Date()
    }
}
